### DISCUSSION TITLE
Extending Cover Entity - Support for multiple motors

### DISCUSSION THREAD

**Discussion by kingy444** [home-assistant/architecture#745] [Entity Models]
Question asked over here https://github.com/home-assistant/core/pull/62788 as part of a merge for hunter Douglas top down bottom up support

while the current method of creating two entities works it is a bit clunky as each motor needs to know where the other is when moving so there is a bit of duplication 

I would like to propose and discuss adding another supported method such as ‘SUPPORT_TILT’ so that these can be implemented in a single entity similar to the below
![8DB51600-E9CC-4030-B65B-C759FF799A25](https://user-images.githubusercontent.com/7148670/156056165-fc176c95-d3b0-471a-85a5-db093c54a7bd.png)

**Comment by kingy444** [comment]
@bdraco  appreciate any help in getting this conversation started
There is at least 1 integration(motion blinds) I believe that would benefit from this type of control and this would work much better with the PowerView top down blinds I am currently trying to implement

while we can create seperate entities it gets a bit messy with each motor needing to know the position of the other before making a move call so this architecture addition would make controlling them much cleaner

**Comment by thecode** [comment]
I am linking to a relevant comment since the description here was not clear to me, would be better to edit it and even post a picture of how it looks like: https://github.com/home-assistant/core/pull/62788#discussion_r811587642

This is what I understand the cover looks like in different states:
![image](https://user-images.githubusercontent.com/1858925/157538051-64c0227e-7417-4fbb-b9ad-84baeb5b8ac7.png)

So questions which are asked on every discussion is how many integrations can fit from such change, do you know about any existing core integrations (or maybe even custom integrations) that would benefit from such feature?

About motor colliding with each other - This *MUST* by protected and handled by the upstream library, if the motors can't be operated separately and need to know about the position of each other the library should handle this and make sure they are not colliding.

**Reply by kingy444** [reply]
The one I immediately know of is Motion Blinds as I came across it with the research on how tdbu was implemented.

I can do some more research if needed but im sure there are other blinds out there that support tdbu that haven’t been implemented as the capability has been a little tedious in implementing

Not wanting the motors to cross is one issue (and obviously important) but the other big issue with the implementation of these as seperate entities is you need to post two values (position1/position2) on every call.

You can submit only one of these, but when you do that the value that is not submitted is treated as 0 resulting in the blind closing the opposite motor and not wanted.  I have provided a copy on the forum to be used and tested under a custom component and this has been one of the very early issues raised.

This isn’t because of the upstream api implementation (same result doing a post in power shell)

**Comment by kingy444** [comment]
Evening Team,

What are next steps here ? As in who OK's moving ahead with something like this ?

The main implication i can see is we would probably need a new service call to move a blind that has 2 motors.

I have a new use case aswell. Until now we had only come across shades that have tilt capability and raise from the bottom (2 capabilities) - There is a type that i am trying to support that raises from the bottom and lowers from the top and tilts (3 capabilities)

With the implentation of TDBU that i have currently been asked to present this wouldnt work for this type of shade

ie - i can currently say and shade that raises from the bottom has tilt capability as that will appear as 1 entity - TDBU currently require 2 entities - so where do you implement the tilt capability in this case ?

I believe this is probably a reason on its own to need to implement DUAL MOTOR blind support

I am happy to work on this but am seeking architecture guidance as advised by @bdraco before making such a change

**Comment by MartinHjelmare** [comment]
> As in who OK's moving ahead with something like this?

There should be consensus among responding, at least two, members. A post should be marked as answer by a member.

**Comment by BradleyFord** [comment]
How many other devices need this support?

It seems a lot of complexity in core for a niche use case

**Reply by kingy444** [reply]
I dont know unfortunately - in my searches the only other two motor blind i came across was motion blinds which appears to integrate in HA via seperate entities at the moment

The issue with HunterDouglas doing it this way is it causes a lot of duplication as the top and bottom motor need to know the position of their alternate when posting to them or it can cause stale data - alternatively i can probably force a refresh post moving these types if that is the architecture direction. Not sure if that is the best approach as it could cause some extra battery drainage but i was already considering this for prevention of sending a command that would close motors too far.

Then - how would you advise handling the use case of a blind with 3 motors ? 
It is reallly nice to have the tilt and raise on one entity so would relly clean it up to have all three on one - but i do get you could consider this niche and it could muddy the waters a bit

are we against a custom service for core integrations as an extension of cover_move ? ie could hunter_douglas imlement its own extension of this service if i went that route for posting two positions to the blind(s) ?

**Reply by Jc2k** [reply]
RE: your stale data / data ownership issue - it's quite common to pull the "state" / fetching logic out of the entity and make it a "view" of an object provided by your support library. So your integration sets up one of the helpers from `homeassistant.helpers.update_coordinator`, and that holds the latest state for the whole device, then multiple entities show their "view" of that data. That avoids duplication of any polling, and stale data, and battery waste.

A more concrete example of this might be a plant sensor that polls water moistness, light levels, temperature and battery level. There are 4 seperate sensors here. Instead of 4 entities that each poll, you'd have a update coordinator that fetched the data once and then 4 entities which just selected their own value from the data the update coordinator holds.

There will still be some code duplication in the cover entities themselves so they can be aware of whether they are a top or a bottom, but i'm sure you could minimize this with the right abstraction.

**Reply by kingy444** [reply]
thanks @Jc2k - i believe we may already be doing that but to be sure could you point me to an integration that does this well so i can compare the code?
I have a feeling we could improve its implementation - is there a way for the integration to push data into the update_coordinator when a user makes a UI based move? ie being two integrations it would be good for the a move intiated from the bottom motor to immediately have the new position loaded into the top motor entity
currently there is a bit of a delay meaning that if a move is made from the top before the bottom motor data has updated then it can result in the bottom motor returning to its previous position - hard to put into words but below is the situation

before any move
* Top Blind Entitity has topMotor 0 and bottomMotor 0
* Bottom Blind Entitity has topMotor 0 and bottomMotor 0

user initiates a move of the bottom blind to open 30%
* Top Blind Entitity has topMotor 0 and bottomMotor 0
* Bottom Blind Entitity has topMotor 0 and bottomMotor 30

user then initiates a move of the top blind to open 15% before the original move has completed
* Top Blind Entitity has topMotor 15 and bottomMotor 0
* Bottom Blind Entitity has topMotor 0 and bottomMotor 30

because the Top Blind had not had time for an update to come through this results in the Bottom Blind closing again
* Top Blind Entitity has topMotor 15 and bottomMotor 0
* Bottom Blind Entitity has topMotor 0 and bottomMotor 0

IF these were 1 entity it wouldnt matter because the data would be up to date immediately and the second move would send both positions in one call and it would set the desired positions
* Top Blind Entitity has topMotor 15 and bottomMotor 30

**Reply by Jc2k** [reply]
The situation you described is exactly what I imagined your problem to be, and one that I don't think it is one that should exist with a shared update coordinator.

Ultimately a coordinator is just a shared object that multiple entities are connected to and that is where the shared state lives. The coordinators know which entities are tied to them and can "wake" them when they become aware of new data, to ensure the HA state machine is kept up to date. So even though the default implementation is just about managing the lifecycle of fetching data, it is reasonably common to use them with event based data as well as polling data. I'd probably start by looking at `DataUpdateCoordinator.async_set_updated_data`. This is used in a few places. One is in `tradfri` to feed observed events into the update coordinator (in addition to polling). Another is `goalzero`. This uses it to feed `async_turn_on` and `async_turn_off` directly into the coordinator. It is up to the code that creates the entity to pass in a shared coordinator, which already hapens for Hunter Douglas Powerview covers (https://github.com/home-assistant/core/blob/dev/homeassistant/components/hunterdouglas_powerview/cover.py#L87).

Not all coordinators act as a live "view" and their entities instead seem to populate properties on `self`. I personally don't like that approach, because the data is stored twice or more, but HA doesn't always follow my head-cannon. One that does it right (IMO) is `AirthingsHeaterEnergySensor.native_value`. `EvilGeniusLight` also has properties that read directly from the coordinator. But ultimately the coordinator should notify all entities to update themselves when there is new data, so this largely doesn't matter if they cache it on the entity as well. 

In your case: even though there are 2 cover entities (one for each motor), both entities should be able to access the exact same underlying raw data from the coordinator. As long as the coordinator is kept up to date there shouldn't be a problem: As soon as one entity learns of an updated motor position, both do.

To be clear I primarily work on homekit_controller. This does not use an update coordinator, mostly because it predates them,  so i'm not the leading expert on this. However the design ended up independently looking a lot like the pattern used by an update coordinator. There the class `HKDevice` plays the roll of the coordinator, and it has a `HKDevice.process_new_events` method that gets called by polling, event notifications and when we call an API on an accessory, including for our cover entity. This means the entity is always up to date with the latest data we have, regardless of the source.

**Reply by kingy444** [reply]
Thanks for the above @Jc2k  - reaching out for some more help now unfortunately

Disclaimer: I am not python expert but understand and can work with most Object Orientated languages so appreciate assistance here

I did try to look at the `DataUpdateCoordinator.async_set_updated_data` as suggested but dont really get it and appreciate if you can point me in the right direction or provide some example of how i would use this in the code.

Im not having a lot of luck - I have tried setting the value of `self.coordinator.data[self._shade.id][ATTR_POSITION_DATA][ATTR_POSITION2]` and the other shade does not immediately update the information (displaying in the UI via extra_state_attributes) but the shade works as expected if i also use `self.coordinator.data[self._shade.id][ATTR_POSITION_DATA][ATTR_POSITION2]` as the source and not the self.attributes

If i call the above in the first 60 seconds of system uptime i get the below - any suggestion on how to deal with this except for simply skipping the step with a `if self.coordinator.data is not None:` ? This prevents the error but ideally you would want to 'PUT' or create rather than throw an error - not sure why it doesnt exist at boot

`2022-05-04 15:07:17 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [139779963669376] 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "/workspaces/core/homeassistant/components/websocket_api/commands.py", line 189, in handle_call_service
    await hass.services.async_call(
  File "/workspaces/core/homeassistant/core.py", line 1627, in async_call
    task.result()
  File "/workspaces/core/homeassistant/core.py", line 1664, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/workspaces/core/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await service.entity_service_call(
  File "/workspaces/core/homeassistant/helpers/service.py", line 677, in entity_service_call
    future.result()  # pop exception if have
  File "/workspaces/core/homeassistant/helpers/entity.py", line 964, in async_request_call
    await coro
  File "/workspaces/core/homeassistant/helpers/service.py", line 714, in _handle_entity_call
    await result
  File "/workspaces/core/homeassistant/components/hunterdouglas_powerview/cover.py", line 236, in async_set_cover_position
    await self._async_move(kwargs[ATTR_POSITION])
  File "/workspaces/core/homeassistant/components/hunterdouglas_powerview/cover.py", line 531, in _async_move
    self.coordinator.data[self._shade.id][ATTR_POSITION_DATA][
TypeError: 'NoneType' object is not subscriptable`

**Reply by kingy444** [reply]
@bdraco - tagging you here incase you can help in the context of this plugin - any idea why self.coordinator.data is empty for the first 60 seconds of the system boot ?

**Reply by Jc2k** [reply]
Sorry I don't have time to help you dig into the code right now, super busy irl.

Without reading the code - if it's none either the first poll hasn't happened yet or it's returning bad data.

If it just hasn't happened yet there are probably 2 things to do:

* the entity should implement the available property (might be misremembering its name) and it should return false until there is some data in the co-ordinator. It won't exercise the rest of your code when this property is returning false so you won't get the exception.
* And there should be a way to make it poll straight away. I vaguely remember that is in the documentation for the coordinator?

**Reply by bdraco** [reply]
The problem with `async_set_updated_data` is it updates the data for all the shades when you only want to chance one.  You need another set of listeners to handle just the single shade being updated.  This discussion doesn't really belong in here though.

**Comment by lilyball** [comment]
Every single blind in my house is a Hunter Douglas PowerView top-down bottom-up blind of this nature. I would really love to have Home Assistant support both the bottom and middle rails in the same object, especially this object can expose a better "range" control for position instead of independent sliders for each rail.

Having said that, as a newcomer to HA, my primary concern is with how this presents in the UI and is exposed to HomeKit. The dashboards seem to be based on individual entities, which suggests that this should be handled by making it one entity with multiple positions. This also makes sense in that the two rails are not independent from each other, and so both rails should always be presented together. It would also work around a potential issue where I issue a service call to open all covers. If the rails are independent entities then both the bottom and middle rails would be asked to open fully, and that's not possible. If both rails are in a single entity then the obvious behavior is to open just the primary shade (i.e. the bottom rail). And it avoids the possibility of assigning the entities to separate areas, or having them not be grouped together on the dashboard.

All of this to say, that while I don't know how common top-down/bottom-up shades are, they're not unheard-of, and there are significant downsides to representing them as independent cover entities. If tilt is a supported attribute on a single cover, then secondary shade position should be as well.

**Comment by pp81381** [comment]
Another use-case for this could be projector screen controllers that drive a motor for the screen and a motor for the mask. At the moment I have created an integration for my controller (a Nice TT6) that creates them as two independent covers. Then I also provide a helper entity that can do things like set them to various aspect ratios. I'm working on a front end widget that can show the net effect of the two covers. Perhaps a "multi-motor" Cover entity would be a neater solution for that?
